File: js/14. Longest Common Prefix.js
/**
 * @param {string[]} strs
 * @return {string}
 */
var longestCommonPrefix = function(strs) {
    let preFixLen = 0;
    let size = strs[0].length;
    for (let i = 0; i < size; i++) {
        if (!checkWithIndex(preFixLen, strs)) {
            break;
        }
        preFixLen++;
    }
    return strs[0].substring(0, preFixLen);
};


var checkWithIndex = function (index, strs) {
    if (index >= strs[0].length) {
        return false;
    }
    let size = strs.length;
    let flag = strs[0][index];
    for (let i = 1; i < size; i++) {
        if (index >= strs[i].length) {
            return false;
        }
        if (flag !== strs[i][index]) {
            return false;
        }
    }
    return true;
};

File: go/14. Longest Common Prefix.go
package main

func checkIndexSame(index int, strs []string) bool {
	if index >= len(strs[0]) {
		return false
	}
	flag := strs[0][index]
	for i := 1; i < len(strs); i++ {
		if len(strs[i]) <= index {
			return false
		}
		if flag != strs[i][index]{
			return false
		}
	}
	return true
}

func longestCommonPrefix(strs []string) string {
	len := len(strs[0])
	preFixIndex := 0
	for i := 0; i < len; i++ {
		if !checkIndexSame(preFixIndex, strs) {
			break
		}
		preFixIndex++
	}
	return strs[0][0:preFixIndex]
}
